### ISSUE TITLE
what to do about `Requires-External` core metadata field?

### ISSUE THREAD

**Issue by rgommers** [rgommers/peps#12]
It turns out there is an existing user, `setuptools-ext` - see https://discuss.python.org/t/pep-725-specifying-external-dependencies-in-pyproject-toml/31888/18. From that discussion:

> [PEP 621 – Storing project metadata in pyproject.toml](https://peps.python.org/pep-0621/) doesn’t currently have anything to say about supplying values for this field, but there are users: [setuptools-ext ](https://github.com/wimglenn/setuptools-ext/)is a build backend which exists mainly to specify this field in pyproject.toml using a tool subsection

Thank you for pointing that out. I am unsure, it’s the first time I hear about this field being used. I’m hoping someone with more experience with core metadata can weigh in about whether this field can and should be reused.

**Comment by rgommers** [comment]
Out of all comments on the PEP so far, this is the one I'm least sure about. On the one hand, the current use of `Requires-External` by `setuptools-ext` looks quite unusual, there are no real semantics as far as I can tell. E.g., in the README exampe, there are two packages (libpng and make) and one name of a programming language (C):
```ini
requires-external = [
    "C",
    "libpng (>=1.5)",
    'make; sys_platform != "win32"'
]
```
Given the lack of prescribed semantics, this only seems to be useful for a build backend and a metadata consumer that explicitly know about each other.

On the other hand, backwards compatibility matters, and it'd be much better to not break anything for `setuptools-ext` users (there don't appear to be many, 53 downloads last month says https://pypistats.org/packages/setuptools-ext). 

@wimglenn how disruptive would it be for you and the users you know of if this field got repurposed?

@pradyunsg do you have any advice here? Are there precedents for this kind of thing?

**Comment by wimglenn** [comment]
The example with C, libpng, make is only due to reusing the existing "official" example of [_Requires-External_](https://packaging.python.org/en/latest/specifications/core-metadata/#requires-external-multiple-use) usage, which has been there for as long as I can remember.  `setuptools-ext` propagates those lines (and some other fields) from the `pyproject.toml` file into the package metadata and hence into the wheel/sdist, since vanilla setuptools can not.

I've never publicized `setuptools-ext` anywhere, and given the nil star count on GitHub I wouldn't be surprised if we're the _only_ user. However, it is used internally at my employer fairly extensively (packaging automation for order of thousands packages). You wouldn't see any downloads of setuptools-ext in PyPI stats for this because we mirror and cache packages we need internally in our own index server.

> this only seems to be useful for a build backend and a metadata consumer that explicitly know about each other

That's quite right, and it's exactly the case here. We also control the host OS/platform, so the external dependencies which may be specified are from a well-defined collection.

> how disruptive would it be for you and the users you know of if this field got repurposed?

It's my expectation that as soon as setuptools implemented support for PEP 725, then setuptools-ext (as a setuptools wrapper) in it's current form would probably just break. If the field got repurposed I'd probably make a new release which cooperates with setuptools again and do something like 

```
[build-system]
requires = ["setuptools-ext > 0.5"]
build-backend = "setuptools_ext"
```

Or, I'd just kill the project and find some other way to do what it was doing, passing around the relevant info without using [core metadata fields](https://packaging.python.org/en/latest/specifications/core-metadata/) at all. It would be mildly disruptive/annoying at $EMPLOYER, but as (probably) the only consumer of the field, not very disruptive in the grand scheme of things.

**Comment by rgommers** [comment]
Thank you @wimglenn, that is very helpful context, much appreciated.

Then it seems like it's at least feasible to reuse the field. We'll see if there are more comments about that choice.

**Comment by rgommers** [comment]
Thinking about it more: this metadata would often be dynamic, since for example vendoring with `auditwheel` would change whether an external runtime requirement exists or not. That would be annoyingly complex to implement correctly in build backends and vendoring tools - just like PEP 770, which after discussion ended up removing any Core Metadata change. See also https://github.com/pypa/auditwheel/pull/577 for how nontrivial it quickly gets.

Perhaps it is better to just deprecate the existing `Requires-External` field but not try to replace it nor actively aim to remove it any time soon (it's more a "don't use that, there's a better way now").

**Comment by jaimergp** [comment]
> Perhaps it is better to just deprecate the existing Requires-External field but not try to replace it nor actively aim to remove it any time soon (it's more a "don't use that, there's a better way now").

What's the "better way" to include external runtime dependencies information in a wheel so it can be used at install time? Not that `pip` or `uv` would know what to do with it, but maybe a [wrapping tool](https://github.com/jaimergp/pyproject-external) could parse those requirements and provide them to the system as part of the wheel install too.

**Comment by rgommers** [comment]
> but it doesn't directly affect `external.optional-dependencies`, which are unlikely to be vendored if they are optional, aren't they?

I don't think that's true. Optional build-time dependencies are not uncommon. See for example Pillow, which IIRC has six or so optional dependencies that if they are used will be linked against and then may or may not be vendored by `auditwheel`.

> What's the "better way" to include external runtime dependencies information in a wheel so it can be used at install time?

Good question. I think I was thinking about build time, at which point one can read `pyproject.toml` directly. At install time, I think it's either adding to Core Metadata or adding a new metadata file (`external.json` I think). It's a really annoying problem to get that right, independent of the format. Thinking about it again, the vendoring business is best left alone, because:
1. Wheels as an intermediate format that are turned into packages for another package manager will not have external dependencies at all; they're just regular dependencies for that package manager.
2. Wheels built locally must have such dependencies installed already, or the build would have failed.
3. Redistributable wheels for PyPI must have them vendored to adhere to the relevant specs (manylinux et al.)

So this leaves us with only needed to express runtime dependencies:

> _If_ we were to include new core-metadata fields, what would it look like? Something like this?

That looks about right.

**Comment by jaimergp** [comment]
> So this leaves us with only needing to express runtime dependencies:
> 
> > If we were to include new core-metadata fields, what would it look like? Something like this?
> 
> That looks about right.

Awesome. I'll add this proposal to #27 if that's ok?

**Comment by rgommers** [comment]
> Awesome. I'll add this proposal to https://github.com/rgommers/peps/pull/27 if that's ok?

Yes please! I was thinking about it some more, and am fairly sure that this rationale is fine. We will need only the required and optional runtime dependencies, not the build-time dependencies nor dependency-groups.
